File: speed/memory/memory.jl
using Printf, Random, Base.Threads

# simple arena allocator struct
mutable struct Arena
    buffer::Vector{UInt8}
    used::Int
    
    Arena(size::Int) = new(Vector{UInt8}(undef, size), 0)
end

function allocate!(arena::Arena, size::Int)::Union{Nothing, Vector{UInt8}}
    # align to 8 bytes
    size = (size + 7) & ~7
    
    if arena.used + size > length(arena.buffer)
        return nothing
    end
    
    start_idx = arena.used + 1
    end_idx = arena.used + size
    arena.used += size
    
    return view(arena.buffer, start_idx:end_idx)
end

function reset!(arena::Arena)
    arena.used = 0
end

# allocation patterns test - sequential, random, producer-consumer
function allocation_patterns_test(iterations::Int)::Float64
    start_time = time_ns()
    
    # sequential allocation pattern
    ptrs = Vector{Vector{UInt8}}(undef, iterations)
    for i in 1:iterations
        size = 64 + ((i - 1) % 256)
        ptrs[i] = Vector{UInt8}(undef, size)
    end
    
    # clear and trigger GC
    ptrs = nothing
    GC.gc()
    
    # random allocation pattern
    Random.seed!(42)
    raw_ptrs = Vector{Vector{UInt8}}(undef, iterations)
    
    for i in 1:iterations
        size = 32 + rand(1:512)
        raw_ptrs[i] = Vector{UInt8}(undef, size)
    end
    
    # random deallocation (shuffle and clear)
    shuffle!(raw_ptrs)
    raw_ptrs = nothing
    GC.gc()
    
    end_time = time_ns()
    _ = iterations  # prevent optimization
    return (end_time - start_time) / 1_000_000
end

# worker function for gc stress test
function gc_stress_worker(thread_id::Int, iterations::Int, counter::Ref{Int})
    Random.seed!(42 + thread_id)
    local_counter = 0
    
    for i in 1:iterations
        size = 16 + rand(1:1024)
        data = Vector{UInt8}(undef, size)
        
        # simulate work
        fill!(data, UInt8(i & 0xFF))
        
        # calculate sum
        total = UInt8(0)
        for j in 1:8:length(data)
            total += data[j]
        end
        _ = total  # prevent optimization
        
        local_counter += 1
    end
    
    atomic_add!(counter, local_counter)
end

# gc stress testing with multiple threads
function gc_stress_test(num_threads::Int, iterations_per_thread::Int)::Float64
    start_time = time_ns()
    
    counter = Ref{Int}(0)
    
    @threads for i in 1:num_threads
        gc_stress_worker(i, iterations_per_thread, counter)
    end
    
    result = counter[]
    _ = result  # prevent optimization
    
    end_time = time_ns()
    return (end_time - start_time) / 1_000_000
end

# cache locality and fragmentation test
function cache_locality_test(iterations::Int)::Float64
    start_time = time_ns()
    
    # allocate small and large objects interleaved
    small_ptrs = Vector{Vector{UInt8}}(undef, iterations)
    large_ptrs = Vector{Vector{UInt8}}(undef, iterations)
    
    Random.seed!(42)
    
    # interleaved allocation pattern
    for i in 1:iterations
        small_size = 16 + rand(1:64)
        large_size = 1024 + rand(1:4096)
        
        small_array = Vector{UInt8}(undef, small_size)
        large_array = Vector{UInt8}(undef, large_size)
        
        # access pattern to test spatial locality
        fill!(small_array, UInt8(i & 0xFF))
        fill!(large_array[1:min(1024, length(large_array))], UInt8((i + 1) & 0xFF))
        
        small_ptrs[i] = small_array
        large_ptrs[i] = large_array
    end
    
    # random access pattern to stress cache
    for i in 1:(iterations ÷ 2)
        idx1 = rand(1:iterations)
        idx2 = rand(1:iterations)
        
        if idx1 <= length(small_ptrs)
            small_array = small_ptrs[idx1]
            total = UInt8(0)
            for j in 1:min(16, length(small_array))
                total += small_array[j]
            end
            _ = total
        end
        
        if idx2 <= length(large_ptrs)
            large_array = large_ptrs[idx2]
            total = UInt8(0)
            for j in 1:64:min(1024, length(large_array))
                total += large_array[j]
            end
            _ = total
        end
    end
    
    end_time = time_ns()
    return (end_time - start_time) / 1_000_000
end

# memory pool performance test
function memory_pool_test(iterations::Int)::Float64
    start_time = time_ns()
    
    # test standard allocation
    std_ptrs = Vector{Vector{UInt8}}(undef, iterations)
    for i in 1:iterations
        data = Vector{UInt8}(undef, 128)
        fill!(data, UInt8(i & 0xFF))
        std_ptrs[i] = data
    end
    
    std_ptrs = nothing
    GC.gc()
    
    # test arena allocation
    arena = Arena(iterations * 128 + 1024)
    arena_ptrs = Vector{Union{Nothing, Vector{UInt8}}}(undef, iterations)
    
    for i in 1:iterations
        ptr = allocate!(arena, 128)
        if ptr !== nothing
            fill!(ptr, UInt8(i & 0xFF))
            arena_ptrs[i] = ptr
        end
    end
    
    # batch deallocation
    reset!(arena)
    
    # test batch allocation
    for batch in 1:10
        for i in 1:(iterations ÷ 10)
            ptr = allocate!(arena, 128)
            if ptr !== nothing
                fill!(ptr, UInt8(i & 0xFF))
            end
        end
        reset!(arena)
    end
    
    end_time = time_ns()
    return (end_time - start_time) / 1_000_000
end

# memory intensive workloads test
function memory_intensive_test(large_size_mb::Int)::Float64
    start_time = time_ns()
    
    size = large_size_mb * 1024 * 1024
    
    # large object allocation
    large_array1 = Vector{UInt8}(undef, size)
    large_array2 = Vector{UInt8}(undef, size)
    
    # memory bandwidth test - sequential write
    @inbounds for i in 1:4096:size
        large_array1[i] = UInt8(i & 0xFF)
    end
    
    # memory copy operations
    copyto!(large_array2, large_array1)
    
    # memory bandwidth test - sequential read
    total = 0
    @inbounds for i in 1:4096:size
        total += Int(large_array2[i])
    end
    _ = total
    
    # memory access pattern test
    Random.seed!(42)
    for i in 1:10000
        offset = rand(1:(size - 64))
        val = large_array1[offset]
        large_array2[offset] = UInt8((Int(val) + 1) & 0xFF)
    end
    
    end_time = time_ns()
    return (end_time - start_time) / 1_000_000
end

function main()
    scale_factor = 1
    
    if length(ARGS) > 0
        try
            scale_factor = parse(Int, ARGS[1])
            if scale_factor <= 0
                scale_factor = 1
            end
        catch
            @warn "Invalid scale factor. Using default 1."
            scale_factor = 1
        end
    end
    
    total_time = 0.0
    
    # warm up Julia JIT compiler
    allocation_patterns_test(100)
    gc_stress_test(2, 100)
    cache_locality_test(100)
    memory_pool_test(100)
    memory_intensive_test(1)
    
    # actual benchmarks
    total_time += allocation_patterns_test(10000 * scale_factor)
    total_time += gc_stress_test(4, 2500 * scale_factor)
    total_time += cache_locality_test(5000 * scale_factor)
    total_time += memory_pool_test(8000 * scale_factor)
    total_time += memory_intensive_test(100 * scale_factor)
    
    @printf "%.3f\n" total_time
end

main()

File: speed/memory/memory.nim
import std/[times, os, strutils, locks]

# simple arena allocator type
type Arena = object
  buffer: seq[byte]
  used: int

proc newArena(size: int): Arena =
  result.buffer = newSeq[byte](size)
  result.used = 0

proc allocate(arena: var Arena, size: int): ptr UncheckedArray[byte] =
  # align to 8 bytes
  let alignedSize = (size + 7) and not 7
  
  if arena.used + alignedSize > arena.buffer.len:
    return nil
  
  result = cast[ptr UncheckedArray[byte]](addr arena.buffer[arena.used])
  arena.used += alignedSize

proc reset(arena: var Arena) =
  arena.used = 0

# get time in milliseconds
proc getTimeMs(): float =
  result = epochTime() * 1000.0

# simple xorshift rng
type XorShift64 = object
  state: uint64

proc newXorShift64(seed: uint64): XorShift64 =
  result.state = seed

proc next(rng: var XorShift64): uint64 =
  rng.state = rng.state xor (rng.state shl 13)
  rng.state = rng.state xor (rng.state shr 7)
  rng.state = rng.state xor (rng.state shl 17)
  result = rng.state

# allocation patterns test - sequential, random, producer-consumer
proc allocationPatternsTest(iterations: int): float =
  let startTime = getTimeMs()
  
  # sequential allocation pattern
  var ptrs = newSeq[seq[byte]](iterations)
  for i in 0..<iterations:
    let size = 64 + (i mod 256)
    ptrs[i] = newSeq[byte](size)
  
  ptrs.setLen(0)
  GC_fullCollect()
  
  # random allocation pattern
  var rng = newXorShift64(42)
  var rawPtrs = newSeq[seq[byte]](iterations)
  
  for i in 0..<iterations:
    let size = 32 + int(rng.next() mod 512)
    rawPtrs[i] = newSeq[byte](size)
  
  # random deallocation (shuffle and clear)
  for i in countdown(rawPtrs.high, 1):
    let j = int(rng.next() mod (i + 1).uint64)
    swap(rawPtrs[i], rawPtrs[j])
  rawPtrs.setLen(0)
  GC_fullCollect()
  
  let finishTime = getTimeMs()
  discard iterations  # prevent optimization
  return finishTime - startTime

# thread data for gc stress test
type ThreadData = object
  threadId: int
  iterations: int
  counter: ptr int
  lock: ptr Lock

proc gcStressWorker(data: ThreadData) {.thread.} =
  var rng = newXorShift64(42.uint64 + data.threadId.uint64)
  
  for i in 0..<data.iterations:
    let size = 16 + int(rng.next() mod 1024)
    var buffer = newSeq[byte](size)
    
    # simulate work
    for j in 0..<size:
      buffer[j] = byte(i and 0xFF)
    
    # calculate sum
    var sum: byte = 0
    for j in countup(0, size - 1, 8):
      sum += buffer[j]
    discard sum  # prevent optimization
    
    acquire(data.lock[])
    inc(data.counter[])
    release(data.lock[])

# gc stress testing with multiple threads
proc gcStressTest(numThreads: int, iterationsPerThread: int): float =
  let startTime = getTimeMs()
  
  var counter = 0
  var lock: Lock
  initLock(lock)
  
  var threads = newSeq[Thread[ThreadData]](numThreads)
  var threadData = newSeq[ThreadData](numThreads)
  
  for i in 0..<numThreads:
    threadData[i] = ThreadData(
      threadId: i,
      iterations: iterationsPerThread,
      counter: addr counter,
      lock: addr lock
    )
    createThread(threads[i], gcStressWorker, threadData[i])
  
  for i in 0..<numThreads:
    joinThread(threads[i])
  
  deinitLock(lock)
  
  let threadResult = counter
  discard threadResult  # prevent optimization
  
  let finishTime = getTimeMs()
  return finishTime - startTime

# cache locality and fragmentation test
proc cacheLocalityTest(iterations: int): float =
  let startTime = getTimeMs()
  
  # allocate small and large objects interleaved
  var smallPtrs = newSeq[seq[byte]](iterations)
  var largePtrs = newSeq[seq[byte]](iterations)
  
  var rng = newXorShift64(42)
  
  # interleaved allocation pattern
  for i in 0..<iterations:
    let smallSize = 16 + int(rng.next() mod 64)
    let largeSize = 1024 + int(rng.next() mod 4096)
    
    smallPtrs[i] = newSeq[byte](smallSize)
    largePtrs[i] = newSeq[byte](largeSize)
    
    # access pattern to test spatial locality
    for j in 0..<smallPtrs[i].len:
      smallPtrs[i][j] = byte(i and 0xFF)
    for j in 0..<min(1024, largePtrs[i].len):
      largePtrs[i][j] = byte((i + 1) and 0xFF)
  
  # random access pattern to stress cache
  for i in 0..<(iterations div 2):
    let idx1 = int(rng.next() mod iterations.uint64)
    let idx2 = int(rng.next() mod iterations.uint64)
    
    if idx1 < smallPtrs.len:
      var sum: byte = 0
      for j in 0..<min(16, smallPtrs[idx1].len):
        sum += smallPtrs[idx1][j]
      discard sum
    
    if idx2 < largePtrs.len:
      var sum: byte = 0
      for j in countup(0, min(1024, largePtrs[idx2].len) - 1, 64):
        sum += largePtrs[idx2][j]
      discard sum
  
  let finishTime = getTimeMs()
  return finishTime - startTime

# memory pool performance test
proc memoryPoolTest(iterations: int): float =
  let startTime = getTimeMs()
  
  # test standard allocation
  var stdPtrs = newSeq[seq[byte]](iterations)
  for i in 0..<iterations:
    stdPtrs[i] = newSeq[byte](128)
    for j in 0..<128:
      stdPtrs[i][j] = byte(i and 0xFF)
  
  stdPtrs.setLen(0)
  GC_fullCollect()
  
  # test arena allocation
  var arena = newArena(iterations * 128 + 1024)
  var arenaPtrs = newSeq[ptr UncheckedArray[byte]](iterations)
  
  for i in 0..<iterations:
    let p = arena.allocate(128)
    if p != nil:
      for j in 0..<128:
        p[j] = byte(i and 0xFF)
      arenaPtrs[i] = p
  
  # batch deallocation
  arena.reset()
  
  # test batch allocation
  for batch in 0..<10:
    for i in 0..<(iterations div 10):
      let p = arena.allocate(128)
      if p != nil:
        for j in 0..<128:
          p[j] = byte(i and 0xFF)
    arena.reset()
  
  let finishTime = getTimeMs()
  return finishTime - startTime

# memory intensive workloads test
proc memoryIntensiveTest(largeSizeMb: int): float =
  let startTime = getTimeMs()
  
  let size = largeSizeMb * 1024 * 1024
  
  # large object allocation
  var largeArray1 = newSeq[byte](size)
  var largeArray2 = newSeq[byte](size)
  
  # memory bandwidth test - sequential write
  for i in countup(0, size - 1, 4096):
    largeArray1[i] = byte(i and 0xFF)
  
  # memory copy operations
  for i in 0..<size:
    largeArray2[i] = largeArray1[i]
  
  # memory bandwidth test - sequential read
  var sum: int64 = 0
  for i in countup(0, size - 1, 4096):
    sum += largeArray2[i].int64
  discard sum
  
  # memory access pattern test
  var rng = newXorShift64(42)
  for i in 0..<10000:
    let offset = int(rng.next() mod (size - 64).uint64)
    let val = largeArray1[offset]
    largeArray2[offset] = byte((val.int + 1) and 0xFF)
  
  let finishTime = getTimeMs()
  return finishTime - startTime

proc main() =
  var scaleFactor = 1
  
  if paramCount() > 0:
    try:
      scaleFactor = parseInt(paramStr(1))
      if scaleFactor <= 0:
        scaleFactor = 1
    except ValueError:
      echo "Invalid scale factor. Using default 1."
      scaleFactor = 1
  
  var totalTime = 0.0
  
  totalTime += allocationPatternsTest(10000 * scaleFactor)
  totalTime += gcStressTest(4, 2500 * scaleFactor)
  totalTime += cacheLocalityTest(5000 * scaleFactor)
  totalTime += memoryPoolTest(8000 * scaleFactor)
  totalTime += memoryIntensiveTest(100 * scaleFactor)
  
  echo totalTime.formatFloat(ffDecimal, 3)

when isMainModule:
  main()